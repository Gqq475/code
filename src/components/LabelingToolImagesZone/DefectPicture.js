import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import classes from './LabelingToolImagesZone.scss'
import { Resizable } from 'react-resizable'
import Draggable from 'react-draggable'
import { throttle } from 'lodash'
import {
  DRAG_DEFECT_MARK,
  RESIZE_DEFECT_MARK
} from 'routes/OJS/modules/OJS'

export class DefectPicture extends React.Component {
  static propTypes = {
    srcOfOriginalImage: PropTypes.string,
    widthOfOriginalImage: PropTypes.number,
    heightOfOriginalImage: PropTypes.number,
    addframe: PropTypes.func,
    handleMarkDrag: PropTypes.func,
    handleMarkResize: PropTypes.func,
    defectList: PropTypes.array,
    currentDefectIndex: PropTypes.number
  }
  render () {
    let {
      srcOfOriginalImage,
      widthOfOriginalImage,
      heightOfOriginalImage,
      addframe,
      handleMarkDrag,
      handleMarkResize,
      defectList
    } = this.props
    return (
      <div
        style={{
          width: `${widthOfOriginalImage}px`,
          height: `${heightOfOriginalImage}px`,
          backgroundImage: `url('${srcOfOriginalImage}')`,
          backgroundSize: 'cover',
          backgroundRepeat: 'no-repeat',
          margin: '0 auto',
          position: 'absolute'
        }}
        onClick={addframe}>
        {defectList.map((val, index) => {
          return (
            <section
              className={`${classes.outerFrame}`}
              key={index.toString()}
              >
              <Draggable
                axis='both'
                position={{
                  x: val.x,
                  y: val.y
                }}
                onDrag={(event, position) => {
                  if (this.props.currentDefectIndex === index) {
                    handleMarkDrag(index, position, this.props.defectList[index])
                  }
                }}
                cancel={'.react-resizable-handle'}
                >
                <Resizable
                  className={
                    (this.props.currentDefectIndex === index)
                    ? `${classes.innerFrame} ${classes['innerFrame--active']}`
                    : `${classes.innerFrame} `
                  }
                  width={val.width}
                  height={val.height}
                  onResize={(event, { element, size }) => {
                    if (this.props.currentDefectIndex === index) {
                      handleMarkResize(index, size, this.props.defectList[index])
                    }
                  }}>
                  <span
                    style={{
                      position: 'absolute',
                      width: val.width + 'px',
                      height: val.height + 'px'
                    }}
                    />
                </Resizable>
              </Draggable>
            </section>
          )
        })
      }
      </div>
    )
  }
}

export default connect(
  (state) => ({
    defectList: state.ojs.defectList,
    currentDefectIndex: state.ojs.currentDefectIndex
  }),
  (dispatch, ownProps) => ({
    handleMarkDrag: throttle((index, newPosition, { width, height }) => {
      let { x, y } = newPosition
      let { widthOfOriginalImage, heightOfOriginalImage } = ownProps
      if (x < -10) {
        x = -10
      } else if (x > (widthOfOriginalImage - width + 10)) {
        x = widthOfOriginalImage - width + 10
      }

      if (y < -10) {
        y = -10
      } else if (y > (heightOfOriginalImage - height + 10)) {
        y = heightOfOriginalImage - height + 10
      }

      dispatch({
        type: DRAG_DEFECT_MARK,
        data: {
          index, x, y
        }
      })
    }, 50),
    handleMarkResize: throttle((index, newSize, { x, y }) => {
      let { width, height } = newSize
      let { widthOfOriginalImage, heightOfOriginalImage } = ownProps
      if (width > widthOfOriginalImage + 10 - x) {
        width = widthOfOriginalImage + 10 - x
      }
      if (height > heightOfOriginalImage + 10 - y) {
        height = heightOfOriginalImage + 10 - y
      }
      dispatch({
        type: RESIZE_DEFECT_MARK,
        data: {
          index, width, height
        }
      })
    }, 50)
  })
)(DefectPicture)
