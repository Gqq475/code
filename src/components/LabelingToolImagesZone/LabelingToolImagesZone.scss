.LabelingToolImagesZone-container {
  width: 60%;
  float: left;
  border: 1px solid #0087DC;
  border-radius: 5px;
  padding: 0 0 10px 0;
  height: 420px;
  margin-top: 20px;
}

.alertMessage {
  display: flex;
  height: 370px;
  align-items: center;
  justify-content: center;
  font-size: 30px;
}

.outerFrame {
  position: relative;
  // set width and height for bounds of innerFrame
  // width: 100%;
  // height: 100%;
}

.innerFrame {
  border: 2px solid rgba(255, 0, 0, 0.3);
  border-radius: 5px;
  top: 0px;
  left: 0px;
  cursor: move;
  &--active {
    border: 3px solid rgba(255, 0, 0, 1);
    z-index: 1;
  }
}

.defectList--active {
  background-color: rgba(16,22,26,.15);
}

.btnGroup {
  margin: 10px;
  display: flex;
  justify-content: flex-end;
  button {
    margin: 0 5px;
  }
}
