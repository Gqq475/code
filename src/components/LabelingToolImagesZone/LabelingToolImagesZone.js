import React from 'react'
import { assign } from 'lodash'
import classes from './LabelingToolImagesZone.scss'
import Magnifier from 'components/Magnifier'
import DefectPicture from './DefectPicture'

type Props = {
  widthOfOriginalImage: Number,
  heightOfOriginalImage: Number,
  defectPictureSource: Function,
  markDefect: Function,
  selectDefect: Function
};

export class LabelingToolImagesZone extends React.Component {
  props: Props;
  constructor (props) {
    super(props)
    this.addframe = this.addframe.bind(this)
    this.enableAddingFrame = this.enableAddingFrame.bind(this)
    this.openMagnifier = this.openMagnifier.bind(this)
    this.closeMagnifier = this.closeMagnifier.bind(this)
    this.selectScaleSize = this.selectScaleSize.bind(this)
    this.state = {
      magnifierIsValid: false,
      addingFrameIsValid: false,
      scaleSize: 2
    }
  }
  addframe (e) {
    if (this.props.defectPictureSource && this.state.addingFrameIsValid) {
      let lengthOfEdge = 20
      let { offsetX, offsetY } = e.nativeEvent
      this.setState({
        addingFrameIsValid: false
      })
      this.props.markDefect({
        width: lengthOfEdge,
        height: lengthOfEdge,
        x: offsetX - lengthOfEdge / 2,
        y: offsetY - lengthOfEdge / 2,
        name: 'Defect Code',
        imageId: this.props.defectPictureSource.id
      })
    }
  }
  enableAddingFrame (event) {
    this.setState(assign({}, this.state, {
      magnifierIsValid: false,
      addingFrameIsValid: true
    }))
  }
  openMagnifier () {
    this.setState(assign({}, this.state, {
      magnifierIsValid: true
    }))
    this.props.selectDefect({ rowIndex: null })
  }
  closeMagnifier () {
    this.setState(assign({}, this.state, {
      magnifierIsValid: false
    }))
  }
  selectScaleSize (e) {
    this.setState(assign({}, this.state, {
      scaleSize: e.target.value * 1
    }))
  }
  render () {
    return (
      <div className={classes['LabelingToolImagesZone-container']}>
        <section className={`${classes.btnGroup}`}>
          <button
            className='pt-button pt-intent-primary pt-icon-add'
            onClick={this.enableAddingFrame}
            >增加标记</button>
          <button
            className='pt-button pt-intent-primary pt-icon-search-template'
            onClick={this.openMagnifier}
            >打开放大镜</button>
        </section>
        <section
          style={{
            width: `${this.props.widthOfOriginalImage}px`,
            margin: '0 auto'
          }}>
          {
            (this.props.defectPictureSource)
            ? (<DefectPicture
              srcOfOriginalImage={this.props.defectPictureSource.path}
              widthOfOriginalImage={this.props.widthOfOriginalImage}
              heightOfOriginalImage={this.props.heightOfOriginalImage}
              addframe={this.addframe}
              />
            )
            : (<span className={`${classes.alertMessage}`}>
              请先选择需要标记的图片
            </span>)
          }
          {
            (this.state.magnifierIsValid) ? (
              <Magnifier
                scaleSize={this.state.scaleSize}
                selectScaleSize={this.selectScaleSize}
                widthOfMagnifier={150}
                srcOfOriginalImage={this.props.defectPictureSource.path}
                widthOfOriginalImage={this.props.widthOfOriginalImage}
                heightOfOriginalImage={this.props.heightOfOriginalImage}
                closeMagnifier={this.closeMagnifier}
                />
            ) : null
          }
        </section>
      </div>
    )
  }
}

export default LabelingToolImagesZone
