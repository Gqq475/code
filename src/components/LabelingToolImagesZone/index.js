import LabelingToolImagesZone from './LabelingToolImagesZone'
export default LabelingToolImagesZone
