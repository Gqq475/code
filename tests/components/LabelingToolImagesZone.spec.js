import React from 'react'
import LabelingToolImagesZone from 'components/LabelingToolImagesZone/LabelingToolImagesZone'
import { DefectPicture } from 'components/LabelingToolImagesZone/DefectPicture'
describe('(Component) LabelingToolImagesZone', () => {
  it('render normally if defectPictureSource is not exist', () => {
    const data = {
      defectPictureSource: 'http://10.120.136.214/1.png'
    }
    const wrapper = shallow(
      <LabelingToolImagesZone {...data} />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
describe('(Component) LabelingToolImagesZone', () => {
  it('render normally if defectPictureSource is exsit', () => {
    const wrapper = shallow(
      <LabelingToolImagesZone />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
describe('(Component) DefectPicture', () => {
  it('render normally', () => {
    const data = {
      'currentDefectIndex': undefined,
      'srcOfOriginalImage': 'http://10.120.136.214/doc/qms/sample_defect_image/TPDPD.jpg',
      'widthOfOriginalImage': 500,
      'defectList': [{
        createTime: 'Jun 7, 2017 6:29:14 AM',
        defectTypeId: 1,
        difficult: 'Difficult',
        height: 25.25,
        id: 24,
        imageId: 1,
        name: 'ImageDefect 05',
        pose: 'POSE 01',
        truncated: 'TRUNCATED 01',
        updateTime: 'Jun 7, 2017 6:29:14 AM',
        width: 20.25,
        x: 10.25,
        y: 25
      }]
    }
    const handleMarkDrag = jest.fn()
    const handleMarkResize = jest.fn()
    const wrapper = shallow(
      <DefectPicture {...data}
        handleMarkDrag={handleMarkDrag}
        handleMarkResize={handleMarkResize}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
