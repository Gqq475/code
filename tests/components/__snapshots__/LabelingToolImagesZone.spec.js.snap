// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) DefectPicture render normally 1`] = `
<div
  style={
    Object {
      "backgroundImage": "url('http://10.120.136.214/doc/qms/sample_defect_image/TPDPD.jpg')",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "cover",
      "height": "undefinedpx",
      "margin": "0 auto",
      "position": "absolute",
      "width": "500px",
    }
  }
>
  <section
    className="outerFrame"
  >
    <Draggable
      allowAnyClick={false}
      axis="both"
      bounds={false}
      cancel=".react-resizable-handle"
      defaultClassName="react-draggable"
      defaultClassNameDragged="react-draggable-dragged"
      defaultClassNameDragging="react-draggable-dragging"
      defaultPosition={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      disabled={false}
      enableUserSelectHack={true}
      grid={null}
      handle={null}
      offsetParent={null}
      onDrag={[Function]}
      onMouseDown={[Function]}
      onStart={[Function]}
      onStop={[Function]}
      position={
        Object {
          "x": 10.25,
          "y": 25,
        }
      }
      transform={null}
    >
      <Resizable
        axis="both"
        className="innerFrame "
        handleSize={
          Array [
            20,
            20,
          ]
        }
        height={25.25}
        lockAspectRatio={false}
        maxConstraints={
          Array [
            Infinity,
            Infinity,
          ]
        }
        minConstraints={
          Array [
            20,
            20,
          ]
        }
        onResize={[Function]}
        width={20.25}
      >
        <span
          style={
            Object {
              "height": "25.25px",
              "position": "absolute",
              "width": "20.25px",
            }
          }
        />
      </Resizable>
    </Draggable>
  </section>
</div>
`;

exports[`(Component) LabelingToolImagesZone render normally if defectPictureSource is exsit 1`] = `
<div
  className="LabelingToolImagesZone-container"
>
  <section
    className="btnGroup"
  >
    <button
      className="pt-button pt-intent-primary pt-icon-add"
      onClick={[Function]}
    >
      增加标记
    </button>
    <button
      className="pt-button pt-intent-primary pt-icon-search-template"
      onClick={[Function]}
    >
      打开放大镜
    </button>
  </section>
  <section
    style={
      Object {
        "margin": "0 auto",
        "width": "undefinedpx",
      }
    }
  >
    <span
      className="alertMessage"
    >
      请先选择需要标记的图片
    </span>
  </section>
</div>
`;

exports[`(Component) LabelingToolImagesZone render normally if defectPictureSource is not exist 1`] = `
<div
  className="LabelingToolImagesZone-container"
>
  <section
    className="btnGroup"
  >
    <button
      className="pt-button pt-intent-primary pt-icon-add"
      onClick={[Function]}
    >
      增加标记
    </button>
    <button
      className="pt-button pt-intent-primary pt-icon-search-template"
      onClick={[Function]}
    >
      打开放大镜
    </button>
  </section>
  <section
    style={
      Object {
        "margin": "0 auto",
        "width": "undefinedpx",
      }
    }
  >
    <Connect(DefectPicture)
      addframe={[Function]}
    />
  </section>
</div>
`;
